h/accounts/views.py

```diff
@@ -234,126 +240,98 @@ class AsyncRegisterController(RegisterController):
         if subscriptions:
-            # Update the subscriptions table
-            subs = json.loads(subscriptions)
-            if username == subs['uri']:
-                s = Subscriptions.get_by_id(self.request, subs['id'])
-                if s:
-                    s.active = subs['active']
-                    self.db.add(s)
-                    return {}
-                else:
-                    return dict(
-                        errors=[
-                            {'subscriptions': _('Non existing subscription')}
-                        ],
-                        code=404
-                    )
-            else:
-                return dict(
-                    errors=[{'username': _('Invalid username')}], code=400
-                )
+            data = json.loads(subscriptions)
+            err = _update_subscription_data(self.request, data)
+            if err is not None:
+                return err
+            return response
 
-        # Password check
-        user = self.User.get_user(self.request, username, pwd)
-        if user:
-            self.request.context = user
-            response = super(ProfileController, self).edit_profile()
+        # Any updates to fields below this point require password validation.
+        #
+        #   `pwd` is the current password
+        #   `password` (used below) is optional, and is the new password
+        #
+        if not User.validate_user(user, appstruct.get('pwd')):
+            return {'errors': [{'pwd': _('Invalid password')}], 'code': 401}
 
-            # Add the user's email into the model dict that eventually gets
-            # returned to the browser. This is needed so that the edit profile
-            # forms can show the value of the user's current email.
-            if self.request.authenticated_userid:
-                user = h.accounts.models.User.get_by_id(
-                    self.request, self.request.authenticated_userid)
-                response.json = {"model": {"email": user.email}}
+        email = appstruct.get('email')
+        if email:
+            email_user = User.get_by_email(self.request, email)
 
-            return response
-        else:
-            return dict(errors=[{'pwd': _('Invalid password')}], code=401)
+            if email_user:
+                if email_user.id != user.id:
+                    return {
+                        'errors': [{'pwd': _('That email is already used')}],
+                    }
 
-    def disable_user(self):
-        request = self.request
-        schema = schemas.EditProfileSchema().bind(request=request)
-        form = deform.Form(schema)
+            response['model']['email'] = user.email = email
 
-        try:
-            appstruct = form.validate(request.POST.items())
-        except deform.ValidationFailure as e:
-            return dict(errors=e.error.children)
+        password = appstruct.get('password')
+        if password:
+            user.password = password
```

Comment:
You don't need to add objects that came from the database in the first place. sqlalchemy does dirty checking. You only need to add new objects.